### PR TITLE
[AST] Allow access-scope based access check inside inaccessible members

### PR DESCRIPTION
**Pull request by xedin**
This aids expression diagnostics to pinpoint problems related to
chaining of inaccessible members  e.g.

```swift
struct A {
  struct B {}
}

_ = A.B()
```

Here both `B` and its synthesized constructor are inaccessible.

### CODE REVIEW

File: lib/AST/Decl.cpp [hunk 1]
```diff
@@ -2824,6 +2824,13 @@ static bool checkAccess(const DeclContext *useDC, const ValueDecl *VD,
   auto access = getAccessLevel();
   auto *sourceDC = VD->getDeclContext();
 
+  // Preserve "fast path" behavior for everything outside of
```
Comment (slavapestov) [inline lib/AST/Decl.cpp]: I think it's the other way around, now you're only using the fast path below for protocol extension members and operators

Comment (xedin) [reply lib/AST/Decl.cpp]: What I understood from @jrose-apple that's how it's supposed to be - preserve "fast path" behavior only for protocols and operators, and everything else is supposed to use scopes?

Comment (xedin) [reply lib/AST/Decl.cpp]: Ah comment!

**Comment by xedin** [comment]
@swift please smoke test compiler performance

**Comment by xedin** [comment]
@swift please smoke test compiler performance

**Comment by jrose-apple** [comment]
Looks good…but what about `isSetterAccessibleFrom`, and your test case that was failing?

**Comment by xedin** [comment]
@jrose-apple Since check has been moved into `checkAccess` it would be picked up by `isSetterAccessibleFrom` right? I'm going to merge my test-case separately, it requires solver changes which I'm currently working on to attempt inaccessible choices.

**Comment by jrose-apple** [comment]
>< Right, sorry. This looks good; if it passes source compat, go for it.

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please smoke test compiler performance

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test compiler performance

**Comment by xedin** [comment]
@swift-ci please test Linux platform

**Comment by jrose-apple** [comment]
@swift-ci Please test Linux

**Comment by xedin** [comment]
@swift-ci please test compiler performance

**Comment by xedin** [comment]
@jrose-apple looks like one metric regressed in debug mode but I don’t know how bad is that since everything else seems to either go down a bit or stayed the same...

**Comment by jrose-apple** [comment]
:-( I think the bot is still not measuring the right things, but the pipe polls one isn't something to worry about. (That one needs a *big* jump to matter.) I think you're good to go!

**Pull request merged**